### PR TITLE
L0 provider: implement min/recommended page size query

### PR DESCRIPTION
**Pull request by igchor**
and allow passing device ordinal through params

### Checklist

- [ ] Code compiles without errors locally
- [ ] All tests pass locally
- [ ] CI workflows execute properly

- [ ] CI workflows, not executed per PR (e.g. Nightly), execute properly 
- [ ] New tests added, especially if they will fail without my changes
- [ ] Added/extended example(s) to cover this functionality
- [ ] Extended the README/documentation
- [ ] All newly added source files have a license
- [ ] All newly added source files are referenced in CMake files
- [ ] Logger (with debug/info/... messages) is used

### CODE REVIEW

File: include/umf/providers/provider_level_zero.h [hunk 1]
```diff
@@ -83,6 +83,14 @@ umf_result_t umfLevelZeroMemoryProviderParamsSetFreePolicy(
     umf_level_zero_memory_provider_params_handle_t hParams,
     umf_level_zero_memory_provider_free_policy_t policy);
 
+/// @brief  Set the device ordinal in the parameters struct.
+/// @param  hParams handle to the parameters of the Level Zero Memory Provider.
+/// @param  deviceOrdinal device ordinal.
+/// @return UMF_RESULT_SUCCESS on success or appropriate error code on failure.
+umf_result_t umfLevelZeroMemoryProviderSetDeviceOrdinal(
```
Comment (bratpiorka) [inline include/umf/providers/provider_level_zero.h:90]: please add this function to def/map files

File: src/utils/utils_level_zero.h [hunk 2]
```diff
@@ -9,6 +9,7 @@
 #define UMF_UTILS_LEVEL_ZERO_H
 
 #include <umf/providers/provider_level_zero.h>
+#include <vector>
```
Comment (bratpiorka) [inline src/utils/utils_level_zero.h:12]: should this be moved below to the __cplusplusplus section?

File: src/provider/provider_level_zero.c [hunk 3]
```diff
@@ -553,6 +481,132 @@ static umf_result_t ze_memory_provider_free(void *provider, void *ptr,
...
+        (umf_level_zero_memory_provider_params_handle_t)params;
+
+    if (!ze_params->level_zero_context_handle) {
+        LOG_ERR("Level Zero context handle is NULL");
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    if ((ze_params->memory_type == UMF_MEMORY_TYPE_HOST) ==
+        (ze_params->level_zero_device_handle != NULL)) {
+        LOG_ERR("Level Zero device handle is NULL");
```
Comment (PatKamin) [inline src/provider/provider_level_zero.c:531]: The log message is incorrect if both conditions are `true`.

File: include/umf/providers/provider_level_zero.h (hunk 1 above)
Comment (vinser52) [inline include/umf/providers/provider_level_zero.h:90]: The function should be renamed `umfLevelZeroMemoryProviderSetDeviceOrdinal` -> `umfLevelZeroMemoryProviderParamsSetDeviceOrdinal`

File: src/utils/utils_level_zero.h (hunk 2 above)
Comment (igchor) [reply src/utils/utils_level_zero.h:12]: No, includes should not be under that ifdef

File: include/umf/providers/provider_level_zero.h (hunk 1 above)
Comment (igchor) [reply include/umf/providers/provider_level_zero.h:90]: Done.

File: src/provider/provider_level_zero.c (hunk 3 above)
Comment (igchor) [reply src/provider/provider_level_zero.c]: Done.

File: src/provider/provider_level_zero.c [hunk 4]
```diff
@@ -569,12 +623,19 @@ static void ze_memory_provider_get_last_native_error(void *provider,
...
+        *pageSize = ze_provider->min_page_size;
+        return UMF_RESULT_SUCCESS;
+    }
+
+    ze_memory_allocation_properties_t properties = {
+        .stype = ZE_STRUCTURE_TYPE_MEMORY_ALLOCATION_PROPERTIES};
+    ze_result_t ze_result = g_ze_ops.zeMemGetAllocProperties(
+        ze_provider->context, ptr, &properties, NULL);
+
+    return ze2umf_result(ze_result);
```
[the remark this answers is not in the document]
Comment (igchor) [reply src/provider/provider_level_zero.c]: Yeah, I didn't test this path (for ptr != nullptr). Done.

File: include/umf/providers/provider_level_zero.h (hunk 1 above)
Comment (igchor) [reply include/umf/providers/provider_level_zero.h]: Done.

File: src/provider/provider_level_zero.c [hunk 5]
```diff
@@ -553,6 +481,133 @@ static umf_result_t ze_memory_provider_free(void *provider, void *ptr,
...
+
+    void *ptr;
+    umf_result_t result = ze_memory_provider_alloc(ze_provider, 1, 0, &ptr);
+    if (result != UMF_RESULT_SUCCESS) {
+        return result;
+    }
+
+    ze_memory_allocation_properties_t properties = {
+        .stype = ZE_STRUCTURE_TYPE_MEMORY_ALLOCATION_PROPERTIES};
+    ze_result_t ze_result = g_ze_ops.zeMemGetAllocProperties(
```
Comment (vinser52) [inline src/provider/provider_level_zero.c:498]: Have you considered `zeVirtualMemQueryPageSize` instead of your current implementation?

Comment (igchor) [reply src/provider/provider_level_zero.c:506]: As far as I know that function is used for zeVirtualMemReserve and zePhysicalMemCreate only. It can theoretically be different than what is used by regular allocations.

File: src/utils/utils_level_zero.h (hunk 2 above)
Comment (igchor) [reply src/utils/utils_level_zero.h]: I didn't notice this was not a c++-only header file. I changed the function to not use any c++ types and just return the number of properties.

File: test/providers/provider_level_zero.cpp [hunk 6]
```diff
@@ -423,6 +435,36 @@ TEST_P(umfLevelZeroProviderTest, levelZeroProviderNullParams) {
     EXPECT_EQ(res, UMF_RESULT_ERROR_INVALID_ARGUMENT);
 }
 
+TEST_P(umfLevelZeroProviderTest, setDeviceOrdinalValid) {
```
Comment (lukaszstolarczuk) [inline test/providers/provider_level_zero.cpp:438]: would be also nice to have a test for invalid case

Comment (igchor) [reply test/providers/provider_level_zero.cpp:438]: Yeah, but there's a problem - it seems L0 does do validate the ordinals right now so even if you pass invalid one, you might succeed.

Comment (lukaszstolarczuk) [reply test/providers/provider_level_zero.cpp:438]: hmm I see, so at least you could add negative test for null params (see testcase above)

File: src/provider/provider_level_zero.c [hunk 7]
```diff
@@ -307,6 +327,18 @@ umf_result_t umfLevelZeroMemoryProviderParamsSetMemoryType(
     return UMF_RESULT_SUCCESS;
 }
 
+umf_result_t umfLevelZeroMemoryProviderParamsSetDeviceOrdinal(
+    umf_level_zero_memory_provider_params_handle_t hParams,
+    uint32_t deviceOrdinal) {
+    if (!hParams) {
+        LOG_ERR("Level zero memory provider params handle is NULL");
```
Comment (lukaszstolarczuk) [inline src/provider/provider_level_zero.c:334]: and a minor thing - pls make `Zero` uppercase here

**Review by vinser52 — APPROVED** [review]

File: test/providers/provider_level_zero.cpp (hunk 6 above)
Comment (igchor) [reply test/providers/provider_level_zero.cpp:438]: Done.

File: src/provider/provider_level_zero.c (hunk 7 above)
Comment (igchor) [reply src/provider/provider_level_zero.c]: Done

Comment (lukaszstolarczuk) [reply src/provider/provider_level_zero.c]: thx!

**Review by lukaszstolarczuk — COMMENTED** [review]
LGTM
